### PR TITLE
cleanup: simplify pickOneNodeForPreemption in preemption

### PR DESCRIPTION
**Pull request by binacs**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Simplify the implementation logic of `pickOneNodeForPreemption` to make it clearer.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by BinacsLee** [comment]
kindly ping @sanposhiho

**Comment by sanposhiho** [comment]
Sorry, I cannot take time for reviewing this soon. I'll definitely keep this in mind and hopefully review it by next weekend. 🙏 

/assign

**Comment by BinacsLee** [comment]
kindly ping @sanposhiho

File: pkg/scheduler/framework/preemption/preemption.go [hunk 1]
```diff
@@ -442,120 +442,85 @@ func pickOneNodeForPreemption(nodesToVictims map[string]*extenderv1.Victims) str
...
+		minHighestPriorityScoreFunc,
+		// There are a few nodes with minimum highest priority victim. Find the
+		// smallest sum of priorities.
+		minSumPrioritiesScoreFunc,
+		// There are a few nodes with minimum highest priority victim and sum of priorities.
+		// Find one with the minimum number of pods.
+		minNumPodsScoreFunc,
+		// There are a few nodes with same number of pods.
+		// Find the node that satisfies latest(earliestStartTime(all highest-priority pods on node))
+		latestStartTimeScoreFunc,
```
Comment (sanposhiho) [inline pkg/scheduler/framework/preemption/preemption.go:501]: Revise the wording if you'd like. (I'm not a good English writer.

```suggestion
	// Each scoreFuncs gives a high score on preferred Node.
	// The functions are executed from the top,
	// and if there are still ties after execution, then we execute next scoreFunc.
	scoreFuncs := []func(string) int64{
		// A node with a minimum number of PDB is preferable.
		minNumPDBViolatingScoreFunc,
		// A node with a minimum highest priority victim is preferable.
		minHighestPriorityScoreFunc,
		// A node with the smallest sum of priorities is preferable.
		minSumPrioritiesScoreFunc,
		// A node with the minimum number of pods is preferable.
		minNumPodsScoreFunc,
		// A node with the latest start time of all highest priority victims is preferable.
		latestStartTimeScoreFunc,
		// If there are still ties, then the first Node in the list is selected.
```

File: pkg/scheduler/framework/preemption/preemption.go [hunk 2]
```diff
@@ -442,120 +442,85 @@ func pickOneNodeForPreemption(nodesToVictims map[string]*extenderv1.Victims) str
...
-		}
-		if numPDBViolatingPods == minNumPDBViolatingPods {
-			minNodes1 = append(minNodes1, node)
-			lenNodes1++
-		}
-	}
-	if lenNodes1 == 1 {
-		return minNodes1[0]
+
+	allCandidates := []string{}
```
Comment (sanposhiho) [inline pkg/scheduler/framework/preemption/preemption.go:446]:
```suggestion
	allCandidates := make([]string, 0, len(nodesToVictims))
```

File: pkg/scheduler/framework/preemption/preemption.go [hunk 3]
```diff
@@ -442,120 +442,85 @@ func pickOneNodeForPreemption(nodesToVictims map[string]*extenderv1.Victims) str
...
-		return minNodes2[0]
-	}
-	nodeToReturn := minNodes2[0]
-	for i := 1; i < lenNodes2; i++ {
-		node := minNodes2[i]
+	latestStartTimeScoreFunc := func(node string) int64 {
 		// Get earliest start time of all pods on the current node.
 		earliestStartTimeOnNode := util.GetEarliestPodStartTime(nodesToVictims[node])
 		if earliestStartTimeOnNode == nil {
 			klog.ErrorS(errors.New("earliestStartTime is nil for node"), "Should not reach here", "node", klog.KRef("", node))
```
Comment (sanposhiho) [inline pkg/scheduler/framework/preemption/preemption.go:481]:
```suggestion
			klog.ErrorS(errors.New("earliestStartTime is nil for node"), "Should not reach here", "node", node.Name)
```

File: pkg/scheduler/framework/preemption/preemption.go (hunk 1 above)
Comment (BinacsLee) [reply pkg/scheduler/framework/preemption/preemption.go]: fixed

File: pkg/scheduler/framework/preemption/preemption.go (hunk 2 above)
Comment (BinacsLee) [reply pkg/scheduler/framework/preemption/preemption.go]: fixed

File: pkg/scheduler/framework/preemption/preemption.go (hunk 3 above)
Comment (BinacsLee) [reply pkg/scheduler/framework/preemption/preemption.go]: Given that node itself is a string (the name of a Node), I think we can just use `node` here.

**Comment by BinacsLee** [comment]
hi @sanposhiho , thanks for review!

I have updated the PR, PTAL

**Comment by BinacsLee** [comment]
ping @sanposhiho

**Review by sanposhiho — APPROVED** [review]
/lgtm

Thanks @BinacsLee. It contributes the readability very much.

**Comment by sanposhiho** [comment]
/assign @damemi

For approval

**Comment by binacs** [comment]
kindly ping @damemi

**Review by damemi — COMMENTED** [review]
/approve
sorry for the delay
cc @Huang-Wei

**Pull request merged**